megatron/model/mamba/mamba.py

```diff
@@ -0,0 +1,360 @@
+import math
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+try:
+    from mamba_ssm.ops.selective_scan_interface import (
+        selective_scan_ref,
+        selective_scan_fn,
+        mamba_inner_fn,
+    )
+    from causal_conv1d import causal_conv1d_fn
+    import einops
+except ModuleNotFoundError:
+    pass
+
+from megatron.model.norms import get_norm
+
+
+# Mamba layer, without parallelism.
+class MambaBlock(nn.Module):
+    def __init__(
+        self,
+        neox_args,
+        init_method,
+        output_layer_init_method,
+    ):
+        super().__init__()
+
+        self.neox_args = neox_args
+
+        dtype = {
+            "fp16": torch.float16,
+            "bf16": torch.bfloat16,
+            "fp32": torch.float32,
+        }[neox_args.precision]
+        self.precision = dtype
+        factory_kwargs = {"device": torch.cuda.current_device(), "dtype": dtype}
+
+        # set variables, mostly following mamba defaults
+        self.d_model = neox_args.hidden_size
+        self.d_state = 16  # state dimensions per channel
+        self.d_conv = 4  # convolution width
+        self.expand = 2  # linear projection expansion factors
+        self.d_inner = int(self.expand * self.d_model)
+        self.dt_rank = math.ceil(self.d_model / 16)  # rank of dt / Delta parameter
+        self.dt_scale = 1.0
+
+        self.dt_init = "random"
+        self.dt_min, self.dt_max, self.dt_init_floor = 0.001, 0.1, 1e-4
+        assert self.dt_init in ["constant", "random"]
+
+        self.use_bias = False  # currently hardcoded to mamba default.
+
+        # up-projection.
+        self.in_proj = nn.Linear(
+            self.d_model,
+            self.d_inner * 2,
+            bias=self.use_bias,
+            **factory_kwargs,
+        )
+        init_method(self.in_proj.weight)
+
+        # convolution.
+        self.conv1d = nn.Conv1d(
+            in_channels=self.d_inner,
+            out_channels=self.d_inner,
+            bias=True,  # hardcode to have bias, currently
+            kernel_size=self.d_conv,
+            groups=self.d_inner,
+            padding=self.d_conv - 1,
+            **factory_kwargs,
+        )
+        # Conv bias sometimes in 32-bit erroneously, when holding other parameters in fp32.
+        # Uncertain why
+        self.conv1d.to(self.precision)
+
+        self.act_fn = F.silu  # we do not allow for
+
+        # x_proj corresponds to s_B(x), s_C(x), s_Delta(x)
+        # in https://arxiv.org/pdf/2312.00752.pdf Algorithm 2
+        # (computes data-dependent B, C, Delta/dt)
+        self.x_proj = nn.Linear(
+            self.d_inner, self.dt_rank + self.d_state * 2, bias=False, **factory_kwargs
+        )
+        init_method(self.x_proj.weight)
+
+        # up-project dt / Delta from dt_rank to d_inner
+        self.dt_proj = nn.Linear(
+            self.dt_rank, self.d_inner, bias=True, **factory_kwargs
```

Comment:
We hardcode the conv biases, but the other projection biases get `self.use_bias`.

Why is this different?

Reply:
No good justification--I can add 2 neox arguments to control the presence of conv, and in/out_proj biases.

Reply:
Updated!

Reply:
I like it, but these two `nn.Linear` biases are still set to `False` and `True` regardless of the new args?

Reply:
the Mamba codebase exposes options in the Mamba block for `bias: bool = False` (controls in_proj and out_proj bias presence, as done here now) and `conv_bias: bool = True` (controls bias presence in conv1d). The other `x_proj` and `dt_proj` biases (False and True, respectively) are hardcoded and there aren't flags for them.

`dt_proj` 's bias is a special case and I believe we should keep it turned on -- Alg. 2 in the Mamba paper defines Delta as `Delta = Tau_{Delta}(Parameter + s_{Delta}(x))` where `s_{Delta}(x) = Broadcast_{D}(Linear_{1}(x))` or as they further explain in section 3.6 can be also `s_{Delta}(x) = Linear_{D}(Linear_{R}(x))` where Linear_R is the delta portion of `x_proj` and Linear_D is the `dt_proj` weight. Then, the `Parameter` term from Alg. 2 can be viewed as the bias term in `dt_proj`, with a special initialization from https://arxiv.org/abs/2206.12037 . 

TL;DR: `dt_proj.bias` is the `Parameter` in Alg. 2 and its presence is assumed.

I'm not aware of anything mentioned in the paper about `x_proj` not being able to have a bias though, and I can't think of a reason it is required to not have it--the fused inner function also seems not to support a bias on it. https://github.com/state-spaces/mamba/blob/34076d664838588a3c97727b263478ab9f621a07/mamba_ssm/ops/selective_scan_interface.py#L164-L167

I've opened an issue to clarify if the hardcoded value is a mistake.

Reply:
Made x_proj bias now also controlled by the same flag!

Reply:
Added this as a comment in case future code-readers have this same confusion.